docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,363 @@
+  // should be viewed as purely informational, and that things could still work.
+  ConditionSeverityInfo ConditionSeverity = "Info"
+
+  // ConditionSeverityNone should apply only if the condition is in state "True".
+  // As ConditionSeverityNone is the default for conditions we use the empty string (coupled with omitempty)
+  ConditionSeverityNone ConditionSeverity = ""
+)
+
+// Conditions define a readiness condition for a Cluster API resource.
+type Condition struct {
+   // Type of condition.
+   // +required
+   Type ConditionType `json:"type" description:"type of status condition"`
+
+   // Status of the condition, one of True, False, Unknown.
+   // +required
+   Status corev1.ConditionStatus `json:"status"`
+
+   // Severity with which to treat failures of this type of condition.
+   // When this is not specified, it defaults to Error.
+   // +optional
+   Severity ConditionSeverity `json:"severity,omitempty"`
+
+   // LastTransitionTime is the last time the condition transitioned from one status to another.
+   // +optional
+   LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+   // The reason for the condition's last transition.
+   // Reasons should be one-word CamelCase.
+   // +optional
+   Reason string `json:"reason,omitempty" description:"one-word CamelCase reason for the condition's last transition"`
+
+   // A human readable message indicating details about the transition.
+   // +optional
+   Message string `json:"message,omitempty" description:"human-readable message indicating details about last transition"`
+}
+
+// Conditions is the schema for the conditions portion of a Cluster API resource.
+type Conditions []Condition
+```
+
+Every Cluster API resource except data-only objects which aren't reconciled (i.e. `KubeadmConfigTemplate`)
+SHOULD have conditions field in the status struct. e.g.
+
+```golang
+// ClusterStatus represent the status for the Cluster object
+type ClusterStatus struct {
+    ...
+    // Conditions define a list of readiness conditions for the Cluster object
+    Conditions Conditions `json:"conditions,omitempty"
+} 
+```
+
+Each Cluster API resource SHOULD define its own set of condition types, e.g. 
+
+```golang
+// ConditionTypes for the cluster object 
+const (
+	ClusterInfrastructureReady ConditionType = "ClusterInfrastructureReady"
+	
+	...
+)
+```
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
+
+Condition types MUST have one of the following suffix:
+
+- `Ready`, for resources which represent an ongoing status, like `ControlplaneReady` or `MachineDeploymentsReady`.
+- `Succeeded`, for resources which run-to-completion, e.g. `CreateVPCSucceeded`
+
+The `Severity` field MUST be set only when `Status=False` and it is designed to provide a standard classification
+of possible conditions failure `Reason`. 
+
+Please note that the combination of `Reason` and `Severity` gives different meaning to a condition failure
```

Comment:
Given my comment above about what a condition is, I wonder if `Severity` might be the wrong word, or even mechanism, for classifying conditions. The information encoded there (i.e. `Info`, `Warning`, `Error`) comes from logging, and even there [some people dispute](https://dave.cheney.net/2015/11/05/lets-talk-about-logging) whether there is a need for more than `Info` and `Error`. Is there a more specific word, which applies to all conditions, which we can use? Or should we just treat Conditions as freeform messages and force callers to know which conditions they care about ahead of time? Could we have a convention for encoding whether a Condition is terminal or not in `Type`?

I would also be curious if this discussion has come up in SIG API Machinery...

Sorry if this note is confusing at all! I probably need to review @bgrant0607's [conditions references](https://github.com/kubernetes/enhancements/pull/1624#issuecomment-603313342) in more detail...

Reply:
> The information encoded there (i.e. Info, Warning, Error) comes from logging

True, but this is also a generally understood way to express:
- You are good
- You should take a look
- You have some problems

> I wonder if Severity might be the wrong word

I have considered `ReasonClass`, `ReasonCategory`, but at the end opted for `Severity` because the focus here is in catching the user's attention.
But I'm more than happy to hear suggestions here...

> should we just treat Conditions as freeform messages and force callers to know which conditions they care about ahead of time

free form message is difficult for consumers, so we are encoding ConditionTypes, Reason, and Severity to make it easier to deal with conditions